### PR TITLE
LF: rationalize archive Parser/Reader/Decoder

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
follow up of #10217

CHANGELOG_BEGIN
CHANGLEOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
cc @oliverse-da this is broken for canton.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala [hunk 1]
```diff
@@ -94,7 +96,6 @@ object GenDarReader {
   }
 }
 
-object DarReader extends GenDarReader[ArchivePayload](is => Try(Reader.readArchive(is)))
-
-object RawDarReader
-    extends GenDarReader[DamlLf.Archive](is => Try(DamlLf.Archive.parseFrom(is.toByteString)))
+object DarParser extends GenDarReader[DamlLf.Archive](Decoder.ArchiveParser)
+object DarReader extends GenDarReader[ArchivePayload](Decoder.ArchiveReader)
+object DarDecoder extends GenDarReader[(PackageId, Ast.Package)](Decoder.ArchiveDecoder)
```
Comment (stefanobaghino-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala:101]: Would these names make it easier to remember which is which?
```suggestion
object DarArchiveReader extends GenDarReader[DamlLf.Archive](Decoder.ArchiveParser)
object DarArchivePayloadReader extends GenDarReader[ArchivePayload](Decoder.ArchiveReader)
object DarPackageAstReader extends GenDarReader[(PackageId, Ast.Package)](Decoder.ArchiveDecoder)
```

**Review by stefanobaghino-da — APPROVED** [review]
One thing, which is probably irrelevant. Otherwise LGTM.

**Comment by oliverse-da** [comment]
> cc @oliverse-da this is broken for canton.

Thank you for the heads-up @remyhaemmerle-da . Are you planning to get the changes in before tomorrow's snapshot? If so it would be great to identify what I will need to change or add the merge changes to https://github.com/DACH-NY/canton/pull/6612

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala:101]: The convention is as follow:
- parsing = proto parsing (i.e. bytes to proto)
- reading = proto parsing + validation (hash and version)
- decoding = proto parsin + validation + proto decoding (proto -> Ast)

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/GenReader.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+package com.daml.lf
+package archive
+
+import com.google.protobuf.{ByteString, CodedInputStream}
+
+import scala.util.Using
+
+final class GenReader[X] private[archive] (val fromCodedInputStream: CodedInputStream => X) {
+
+  // close the stream unconditionally
```
Comment (cocreature) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/GenReader.scala:13]: is that deliberate? as mentioned before I generally find code easier to reason about if the code responsible for opening a stream is responsible for closing it as well.

**Review by cocreature — APPROVED** [review]
Nice!

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/GenReader.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/GenReader.scala:13]: Following the comment on the other PR, I removed the `close`, but forgot the comment.

**Review by miklos-da — APPROVED** [review]

**Pull request merged**